### PR TITLE
[Feature] Support vote fusion

### PR DESCRIPTION
**Pull request by THU17cyz**
Added the module which fuses 2d and 3d and generates 3 cues:

-  Semantic
-  Texture
- Geometric

### CODE REVIEW

**Comment by THU17cyz** [comment]
Just updated as requested:

- Coords transformation during fusion as a separated module called Coord3DTransformation
- Replaced original code in vote fusion and point fusion with the module above
- Added unit test for this module
- Also updated unit test for Coord3DMode convert_point method, including a bug fix, and added TODO in that method because LIDAR implementation is still not correct
- Modified some docstring and comments

File: mmdet3d/models/fusion_layers/coord_transform.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+            Whether flip point cloud along vertical direction.
+    """
+
+    def __init__(self,
+                 dtype,
+                 device,
+                 coords_type,
+                 img_meta=None,
+                 pcd_rotate_mat=None,
+                 pcd_scale_factor=None,
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/coord_transform.py:41]: Should use the default value for this.
`pcd_scale_factor=1.0, pcd_trans_factor=0, pcd_horizontal_flip=False, pcd_vertical_flip=False`.

**Comment by ZwwWayne** [comment]
Unit test failed.

File: mmdet3d/core/bbox/structures/coord_3d_mode.py [hunk 2]
```diff
@@ -245,7 +253,7 @@ def convert_point(point, src, dst, rt_mat=None):
         else:
             xyz = arr[:, :3] @ rt_mat.t()
 
-        remains = arr[..., 3:]
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/structures/coord_3d_mode.py:248]: Is this change necessary?

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                if 'scale_factor' in img_meta.keys() else [1.0, 1.0])
+
+            img_flip = img_meta['flip'] if 'flip' in \
+                img_meta.keys() else False
+            img_crop_offset = (
+                seed_3d_depth.new_tensor(img_meta['img_crop_offset'])
+                if 'img_crop_offset' in img_meta.keys() else 0)
+
+            # first reverse the data transformations
+            xyz_depth = coords_trans(seeds_3d_depth, 'HTSR', reverse=True)
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/vote_fusion.py:77]: This is a little bit hard code. If necessary, the img_meta should contain the information of 'HTSR'

File: mmdet3d/core/bbox/structures/coord_3d_mode.py (hunk 2 above)
Comment (THU17cyz) [reply mmdet3d/core/bbox/structures/coord_3d_mode.py:248]: This is not in accordance with the others, so I unified it.

Comment (THU17cyz) [reply mmdet3d/core/bbox/structures/coord_3d_mode.py:248]: if we want to support any shape of input pc, this function needs to be refactored

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        """
+        img_features = []
+        masks = []
+        for i, data in enumerate(
+                zip(imgs, bboxes_2d_rescaled, seeds_3d_depth, img_metas)):
+            img, bbox_2d_rescaled, seed_3d_depth, img_meta = data
+            bbox_num = bbox_2d_rescaled.shape[0]
+            seed_num = seed_3d_depth.shape[0]
+
+            # xyz_depth = seed_3d_depth.clone()
```
Comment (Tai-Wang) [inline mmdet3d/models/fusion_layers/vote_fusion.py:53]: remove this unused code?

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+                mask = mask.floor().int()
+                mask = mask.transpose(1, 0).reshape(-1).bool()
+
+            # clear the padding
+            img = img[:, :img_shape[0], :img_shape[1]]
+            img_flatten = img.reshape(3, -1).float()
+            img_flatten += torch.tensor(
+                self.img_norm_cfg['mean'], device=img.device).unsqueeze(-1)
+            img_flatten = img_flatten - 128.0
+            img_flatten = img_flatten / 255.0
```
Comment (Tai-Wang) [inline mmdet3d/models/fusion_layers/vote_fusion.py:232]: The normalization seems a little hard? If it is a must, at least we can merge these two lines.

File: mmdet3d/models/fusion_layers/coord_transform.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    # "R" stands for rotation;
+    # "H" stands for horizontal flip;
+    # "V" stands for vertical flip.
+    pipeline_mapping = {
+        'T': translate_func,
+        'S': scale_func,
+        'R': rotate_func,
+        'H': horizontal_flip_func,
+        'V': vertical_flip_func
+    }
```
Comment (Tai-Wang) [inline mmdet3d/models/fusion_layers/coord_transform.py:82]: The pipeline mapping is still like hard code, maybe at least we need to do some assertion/make it more compatible with any input (e.g. if op in pipeline does not belong to current supported pipeline_mapping?)

**Review by Tai-Wang — COMMENTED** [review]
`transformation_3d_pipeline` is still not much elegant. Is there any other workaround? If not, we can make it more compatible and safe in the related implementation.

File: mmdet3d/models/fusion_layers/vote_fusion.py (hunk 5 above)
Comment (THU17cyz) [reply mmdet3d/models/fusion_layers/vote_fusion.py]: Merged the last two lines. This is to keep the same as the official implementation

File: mmdet3d/models/fusion_layers/coord_transform.py (hunk 6 above)
Comment (THU17cyz) [reply mmdet3d/models/fusion_layers/coord_transform.py:82]: I added an assertion currently since making it compatible with any input seems not very necessary and last time Wenwei suggested not to over-engineer. Maybe you can take a look? @ZwwWayne

**Comment by THU17cyz** [comment]
Updated according to the review.

File: mmdet3d/core/bbox/structures/utils.py [hunk 7]
```diff
@@ -122,7 +122,17 @@ def points_cam2img(points_3d, proj_mat):
         torch.Tensor: Points in image coordinates with shape [N, 2].
     """
     points_num = list(points_3d.shape)[:-1]
+
     points_shape = np.concatenate([points_num, [1]], axis=0).tolist()
+    d1, d2 = proj_mat.shape[:2]
+    assert (d1 == 3 and d2 == 3) or (d1 == 3 and d2 == 4) or (d1 == 4
+                                                              and d2 == 4)
```
Comment (Tai-Wang) [inline mmdet3d/core/bbox/structures/utils.py:129]: Please add instruction to the assertion like assert ..., "The shape of projection matrix is not supported."

File: mmdet3d/models/fusion_layers/coord_transform.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    for op in list(pipeline):
+        assert op in pipeline_mapping.keys()
```
Comment (Tai-Wang) [inline mmdet3d/models/fusion_layers/coord_transform.py:84]: Please also add assertion information.

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+                mask = mask.floor().int()
+                mask = mask.transpose(1, 0).reshape(-1).bool()
+
+            # clear the padding
+            img = img[:, :img_shape[0], :img_shape[1]]
+            img_flatten = img.reshape(3, -1).float()
+
+            # normalization
+            # in accordance with the official implementation
+            img_flatten += torch.tensor(
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/vote_fusion.py:231]: The image should already been normalized after loading?

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+                self.img_norm_cfg['mean'], device=img.device).unsqueeze(-1)
+            img_flatten = (img_flatten - 128.0) / 255.0
+
+            # take the normalized pixel value as texture cue
+            uv_flatten = uv_rescaled[:, 1].round().long() * \
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/vote_fusion.py:236]: Do they need to be long or int tensor after round() or floor() operation?

File: mmdet3d/models/fusion_layers/vote_fusion.py (hunk 9 above)
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/vote_fusion.py:231]: may use image.new_tensor()

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+                delta_u = delta_u * z_cam / calibs['K'][i, 0, 0]
+                delta_v = delta_v * z_cam / calibs['K'][i, 0, 0]
+
+                imvote = torch.cat(
+                    [delta_u, delta_v,
+                     torch.zeros_like(delta_v)], dim=-1).reshape(-1, 3)
+
+                # convert from camera coords to depth coords
+                imvote = Coord3DMode.convert_point(
+                    imvote.view((-1, 3)).double(),
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/vote_fusion.py:160]: double or float?

File: tests/test_utils/test_coord_3d_mode.py [hunk 12]
```diff
@@ -62,21 +62,21 @@ def test_points_conversion():
 
     convert_depth_points = cam_points.convert_to(Coord3DMode.DEPTH)
     expected_tensor = torch.tensor([[
-        -5.2422e+00, -2.9757e-01, 4.0021e+01, 6.6660e-01, 1.9560e-01,
```
Comment (ZwwWayne) [inline tests/test_utils/test_coord_3d_mode.py:65]: Ensure the correctness as this changes the expected_tensor.

**Review by ZwwWayne — APPROVED** [review]

File: mmdet3d/models/fusion_layers/vote_fusion.py (hunk 9 above)
Comment (THU17cyz) [reply mmdet3d/models/fusion_layers/vote_fusion.py:231]: This is only to be in accordance with the official implementation, which subtracts 128 from the original image and divides it by 255.

Comment (THU17cyz) [reply mmdet3d/models/fusion_layers/vote_fusion.py:231]: The normalization used in 2d detection should make no difference. Shall I use this instead?

File: mmdet3d/models/fusion_layers/vote_fusion.py (hunk 10 above)
Comment (THU17cyz) [reply mmdet3d/models/fusion_layers/vote_fusion.py]: It must be long or the following op will fail

**Comment by THU17cyz** [comment]
@Tai-Wang @ZwwWayne I have just fixed all the issues mentioned before. Additionally, I added a unittest for vote fusion to make sure the rewrote VoteFusion is correct.

File: mmdet3d/core/bbox/structures/utils.py [hunk 13]
```diff
@@ -122,7 +122,20 @@ def points_cam2img(points_3d, proj_mat):
         torch.Tensor: Points in image coordinates with shape [N, 2].
     """
     points_num = list(points_3d.shape)[:-1]
+
     points_shape = np.concatenate([points_num, [1]], axis=0).tolist()
+    assert len(proj_mat.shape) == 2, 'The dimension of the projection'\
+        'matrix should be 2 instead of %d.' % len(proj_mat.shape)
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/structures/utils.py:128]: Use f-string

File: mmdet3d/core/bbox/structures/utils.py [hunk 14]
```diff
@@ -122,7 +122,20 @@ def points_cam2img(points_3d, proj_mat):
...
+    d1, d2 = proj_mat.shape[:2]
+    assert (d1 == 3 and d2 == 3) or (d1 == 3 and d2 == 4) or (
+        d1 == 4 and d2 == 4), 'The shape of the projection matrix'\
+        ' (%d*%d) is not supported.' % (d1, d2)
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/structures/utils.py:132]: Use f-string

File: mmdet3d/datasets/pipelines/transforms_3d.py [hunk 15]
```diff
@@ -96,10 +96,15 @@ def __call__(self, input_dict):
                 ) < self.flip_ratio_bev_vertical else False
                 input_dict['pcd_vertical_flip'] = flip_vertical
 
+        if 'transformation_3d_pipeline' not in input_dict.keys():
```
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/transforms_3d.py:99]: May use `transformation_3d_flow`

File: mmdet3d/datasets/pipelines/transforms_3d.py [hunk 16]
```diff
@@ -96,10 +96,15 @@ def __call__(self, input_dict):
...
+            input_dict['transformation_3d_pipeline'] = ''
+
         if input_dict['pcd_horizontal_flip']:
             self.random_flip_data_3d(input_dict, 'horizontal')
+            input_dict['transformation_3d_pipeline'] += 'H'
```
Comment (ZwwWayne) [inline mmdet3d/datasets/pipelines/transforms_3d.py:104]: Would list will be better than string?

File: mmdet3d/models/fusion_layers/coord_transform.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+    pipeline_mapping = {
+        'T': translate_func,
+        'S': scale_func,
+        'R': rotate_func,
+        'H': horizontal_flip_func,
+        'V': vertical_flip_func
+    }
+    for op in list(pipeline):
+        assert op in pipeline_mapping.keys(), 'This data '\
+            'transformation op (%s) is not supported' % op
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/coord_transform.py:85]: Use f-string

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 18]
```diff
@@ -0,0 +1,240 @@
...
+            # origin bboxes
+            img_h, img_w, _ = img_shape
+            ori_h, ori_w, _ = ori_shape
+
+            img_scale_factor = (
+                seed_3d_depth.new_tensor(img_meta['scale_factor'][:2])
+                if 'scale_factor' in img_meta.keys() else [1.0, 1.0])
+
+            img_flip = img_meta['flip'] if 'flip' in \
+                img_meta.keys() else False
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/vote_fusion.py:59]: we can simply use `img_meta` rather than `img_meta.keys()`

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 19]
```diff
@@ -0,0 +1,240 @@
...
+            bbox_2d_origin = bbox_2d_rescaled.clone().float()
+            img_shape = img_meta['img_shape']
+            ori_shape = img_meta['ori_shape']
+            # origin bboxes
+            img_h, img_w, _ = img_shape
+            ori_h, ori_w, _ = ori_shape
+
+            img_scale_factor = (
+                seed_3d_depth.new_tensor(img_meta['scale_factor'][:2])
+                if 'scale_factor' in img_meta.keys() else [1.0, 1.0])
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/vote_fusion.py:56]: img_meta.keys() -> img_meta

File: mmdet3d/models/fusion_layers/vote_fusion.py [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+                # apply transformation to lifted imvotes
+                imvote = apply_3d_transformation(
+                    imvote, 'DEPTH', img_meta, reverse=False)
+
+                seed_3d_expanded = seed_3d_expanded.reshape(imvote.shape)
+
+                # ray angle
+                ray_angle = seed_3d_expanded + imvote
+                ray_angle /= torch.sqrt(torch.sum(ray_angle**2, -1) +
+                                        1e-6).unsqueeze(-1)
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/vote_fusion.py:170]: May create a variable named EPS = 1e-6, and use EPS in these places.

File: mmdet3d/models/fusion_layers/coord_transform.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+        rotate_func = partial(pcd.rotate, rotation=pcd_rotate_mat.inverse())
+
+        # reverse the pipeline
+        pipeline = pipeline[::-1]
+    else:
+        scale_func = partial(pcd.scale, scale_factor=pcd_scale_factor)
+        translate_func = partial(pcd.translate, trans_vector=pcd_trans_factor)
+        rotate_func = partial(pcd.rotate, rotation=pcd_rotate_mat)
+
+    # "T" stands for translation;
```
Comment (ZwwWayne) [inline mmdet3d/models/fusion_layers/coord_transform.py:71]: This note should be put in the Note section in the docstring

**Review by ZwwWayne — APPROVED** [review]
